### ISSUE TITLE
ASQ version 7 fails as MessageMetadataRegistry is missing from settings

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/ServiceControl.TransportAdapter#23]
When `TransportAdapterConfig` is created, ASQ throws an exception as it [requires `MessageMetadataRegistry` to be found in the settings](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/88f6b663929ec44ce2ed5d1088ba5b9df130686b/src/Transport/AzureStorageQueueTransport.cs#L35) to register `MessageWrapper` needed by ASQ transport to work.

When transport is created outside of an NSB endpoint context, MessageMetadataRegistry is missing.

```
System.Collections.Generic.KeyNotFoundException
  HResult=0x80131577
  Message=The given key (NServiceBus.Unicast.Messages.MessageMetadataRegistry) was not present in the dictionary.
  Source=NServiceBus.Core
  StackTrace:
   at NServiceBus.Settings.SettingsHolder.Get(String key)
   at NServiceBus.Settings.SettingsHolder.Get[T]()
   at NServiceBus.AzureStorageQueueTransport.Initialize(SettingsHolder settings, String connectionString)
   at NServiceBus.Raw.InitializableRawEndpoint.<Initialize>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Raw.RawEndpoint.<Start>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.TransportAdapter.AuditForwarder`2.<Start>d__3.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.TransportAdapter.ServiceControlTransportAdapter`2.<Start>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Program.<Main>d__0.MoveNext() in C:\github\nsb\docs.particular.net\samples\servicecontrol\adapter-asb-multi-namespace\SCTransportAdapter_1\Adapter\Program.cs:line 69
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Program.<Main>()
```

**ASQ affected versions**
- Version 7
- Version 8

**Comment by SeanFeldman** [comment]
The assumption that SC.TransportAdapter is making is that transport message pump does not require Core's serialization and can send/receive messages without it. With ASQ this breaks down as ASQ reaches into Core's default serializer to serialize/deserialize transport messages upon sending/receiving operations.

Default Core's serializer is not enough though. If SC endpoint is running on the default XML serialization and an adopted endpoint is running with another serialization, both need to be configured for SC.TransportAdapter.

**Comment by SzymonPobiega** [comment]
@SeanFeldman here's the code that needs to be added so that ASQ works: https://github.com/SzymonPobiega/NServiceBus.Raw/blob/release-3.0/src/AcceptanceTests.AzureStorageQueues/When_sending_to_another_endpoint_ASQ.cs. I know it is super ugly but there is nothing I can do from NSB.Raw perspective to fix it.

**Comment by SeanFeldman** [comment]
The code will help with registering the default serializer and `MessageMetadataRegistry`. Where would we get the settings from @SzymonPobiega? `TransportAdapterConfig` expects to have these things, so somehow this needs to be either passed via `TransportAdapterConfig`'s constructor as a callback or another option that I don't know.

**Comment by SeanFeldman** [comment]
To be more specific (using ASB sample): https://github.com/Particular/docs.particular.net/blob/db6d607e156166dcc41a01b786949a26f6b2d69d/samples/servicecontrol/adapter-asb-multi-namespace/SCTransportAdapter_1/Adapter/Program.cs#L13

**Comment by SzymonPobiega** [comment]
The method `SetupTransport(TransportExtensions<AzureStorageQueueTransport> extensions)` from the test I've linked would passed to `CustomizeEndpointTransport` and/or `ConfigureServiceControlTransport` of the adapter config.

I can prepare a TA ASQ sample but not sure what thing we should show there? What is the most common thing people would use with ASQ but can't because of SC's limitations?

**Comment by SeanFeldman** [comment]
> The method SetupTransport(TransportExtensions<AzureStorageQueueTransport> extensions) from the test I've linked would passed to CustomizeEndpointTransport and/or ConfigureServiceControlTransport of the adapter config.

Ah, you're referring to something that _would happen_, not something that is already in place. Yes, then it would work. As long as there's a way to inject default serializer prior to transport being created.

Don't think we need a Transport Adapter Sample for ASQ. What we do need is to document the fact that serializer is required and add that to Service Control documentation.

**Comment by SeanFeldman** [comment]
@SzymonPobiega what you pointed me to was for ASQ v8, this issue was for v7 😉 
The code required for v7 is slightly different:

```c#
bool isMessageType(Type t) => t == typeof(MessageWrapper);
var conventions = new Conventions();                  // required for NSB v6
conventions.AddSystemMessagesConventions(isMessageType);

var ctor = typeof(MessageMetadataRegistry).GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, 
   null, new[] { typeof(Conventions) }, null);
settings.Set<MessageMetadataRegistry>(ctor.Invoke(new object[] { conventions }));
```

**Comment by SeanFeldman** [comment]
Exception is still thrown
```
System.Collections.Generic.KeyNotFoundException
  HResult=0x80131577
  Message=The given key (NServiceBus.Unicast.Messages.MessageMetadataRegistry) was not present in the dictionary.
  Source=NServiceBus.Core
  StackTrace:
   at NServiceBus.Settings.SettingsHolder.Get(String key)
   at NServiceBus.Settings.SettingsHolder.Get[T]()
   at NServiceBus.AzureStorageQueueTransport.Initialize(SettingsHolder settings, String connectionString)
   at NServiceBus.Raw.InitializableRawEndpoint.<Initialize>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Raw.RawEndpoint.<Start>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.TransportAdapter.AuditForwarder`2.<Start>d__3.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.TransportAdapter.ServiceControlTransportAdapter`2.<Start>d__1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Program.<Main>d__0.MoveNext() in C:\github\nsb\docs.particular.net\samples\servicecontrol\adapter-asb-multi-namespace\SCTransportAdapter_1\Adapter\Program.cs:line 90
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Program.<Main>()
```

**Comment by SeanFeldman** [comment]
The following workarounds get the adapter endpoint to start:

```c#
var settings = transport.GetSettings();
/*// Register serializer used to serialize MessageWrapper (custom MessageWrapper serializer or endpoint's serializer different than JSON)
var serializer = Tuple.Create(new XmlSerializer() as SerializationDefinition, new SettingsHolder());
settings.Set("MainSerializer", serializer);*/

var conventions = new Conventions();
settings.Set<Conventions>(conventions);

var ctor = typeof(MessageMetadataRegistry).GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, null, new[] { typeof(Conventions) }, null);
var registry = (MessageMetadataRegistry)ctor.Invoke(new object[] { conventions });
settings.Set<MessageMetadataRegistry>(registry);
```

@SzymonPobiega this is somewhat more simplified as the `TransportDefinition.Initialize()` only cares to find `MessageMetadataRegistry` and `Conventions` in the `SettingsHolder`, and handles `MessageWrapper` registration itself.

For serializer (code commented out), ASQ adapter [cannot set JsonSerilizer blindly](https://github.com/SzymonPobiega/NServiceBus.Raw/blob/aeafee6f893c60e88b0c2c6b832691789a410df6/src/AcceptanceTests.AzureStorageQueues/When_sending_to_another_endpoint_ASQ.cs#L21-L22). It has to use serializer used to serialize message wrapper (either a serializer registered for an endpoint or serializer selected for message wrapper via [configuration API](https://docs.particular.net/transports/azure-storage-queues/configuration#custom-envelope-unwrapper)).

**Comment by SzymonPobiega** [comment]
Wow, this is really messy :-( I agree that we can't really tell customers to do it...

**Comment by danielmarbach** [comment]
@SeanFeldman @SzymonPobiega https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/304 would remove the dependency to the conventions and the `MessageMetadataRegistry`. @SzymonPobiega you told me that the adapter will use the latest released version so releasing patch 7.5.2 would make things simpler and we would be left only with the serializer hack for now.

**Comment by SeanFeldman** [comment]
Released ASQ 7.5.2 has removed the need to workaround `MessageMetadataRegistry` and `Conventions` (yay!), but an `System.NullReferenceException` is thrown upon sending related to serialization.

<details> <summary>Stacktrace</summary>

System.NullReferenceException: 'Object reference not set to an instance of an object.'

   at NServiceBus.AzureStorageQueues.Dispatcher.<Send>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.MutateOutgoingTransportMessageBehavior.<InvokeOutgoingTransportMessageMutators>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.SerializeMessageConnector.<Invoke>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.MutateOutgoingMessageBehavior.<InvokeOutgoingMessageMutators>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.UnicastSendRouterConnector.<Invoke>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Program.<Main>d__0.MoveNext() in C:\github\nsb\docs.particular.net\samples\servicecontrol\adapter-asb-multi-namespace\SCTransportAdapter_1\Sales\Program.cs:line 89
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Program.<Main>()

</details>

Will keep this issue opened until can verify what's causing it and if the workaround for the serializer is a suitable one as dispatcher also uses serializer.

**Comment by SeanFeldman** [comment]
Good news. Adapter is working.
Not so great news - there's a [bug with ASQ](https://github.com/Particular/NServiceBus.AzureStorageQueues/issues/308) that causes adapter not to work (technically endpoints) when secured connections are used.
News to break it even - there's a workaround for the bug, so ASQ can be used with SC transport adapter.

Closing this issue as resolved.
